how-we-work/how-to-upgrade-dependencies.md

```diff
@@ -0,0 +1,75 @@
+# How to Upgrade Dependencies
+
+Once a month at the beginning of the month, the oncall engineer should visit the current versions of package dependencies that we have for both the frontend (`npm`) and the backend (`pip`) for `corpora-data-portal` and `cellxgene`. We do this so that we keep up to date with any security vulnerabilities that may exist in older versions and so that we don't end up too out-of-date with improvements that occur in our dependencies causing forward-incompatibility. The sections below describe the policies and instructions for each subset of our product suite.
+
+As a general rule please do not directly import the latest version of the downstream package that contains the vulnerability. For example, if we import package `A` and package `A` depends on package `B` which depends on package `C` which contains a vulnerability, please do not directly import the latest version of package `C` directly into our package. If a newer version of package `A` exists where they have addressed the vulnerability, directly upgrade `A`. Otherwise, wait until there is a new one available.
+
+For an overview of current dependency issues we have in cellxgene Explorer and cellxgene Portal, you can check the two corresponding projects in [Snyk](https://app.snyk.io/org/cellxgene). Note that you should take the findings with a grain of salt as some findings may not have remediation and therefore may continue to exist after everything else has been upgraded. It's a good place to double check your work though.
+
+## Upgrading cellxgene Explorer
+
+### Upgrading cellxgene Explorer frontend
+
+There are two commands to run to figure out which npm packages to upgrade:
+
+`npm audit` -- this command generates a report of known vulnerabilities in dependencies described in `cellxgene/client/package.json`.
+
+`npm outdated` -- this command generates a report of all dependencies in `cellxgene/client/package.json` and whether there are newer versions available.
+
+Run these two commands and upgrade packages to the latest available. Please also make any necessary changes in the client code to ensure that that the frontend still works as expected and is not using outdated references. Finally, run `npm i` to install the packages and update the corresponding `package-lock.json`.
```

Comment:
I'm thinking upgrading a library might need a little more guideline for different levels of upgrade: major, minor, patch?

E.g.,
```md
...

`npm outdated` -- this command generates a report of all dependencies in `cellxgene/client/package.json` and whether there are newer versions available.

Semantic version: Major.Minor.Patch (e.g., v1.2.3 means Major 1, Minor 2, Patch 3)

For major upgrade: Check the package's changelog, especially the breaking changes section first to see if such upgrade is straight forward to complete without significant effort. If so, upgrade away and make sure the affected functionalities still work. If significant effort is required, please create a ticket, document the required steps for upgrade, and bring it to the team's attention for pointing and prioritization

For minor and patch upgrade: In theory there should not be any breaking changes, so just upgrading the packages should not break anything. HOWEVER, not all packages follow semantic versioning, so please notice what packages have been upgraded in `package.json` and test their related functionality accordingly to ensure everything is fine
```

Also we don't need to manually update `package-lock.json` since it's machine generated file 😄 !

Reply:
Do we want to recommend re-building the package lock on semvar major upgrades?

Also, we might want to shout out helper tools like [npm-check-updates](https://github.com/raineorshine/npm-check-updates) and [npm-check](https://github.com/dylang/npm-check)

Reply:
ooooh TIL! `npm-check` seems to provide more functionality, should we recommend that over `npm-check-updates`?

By re-building `package-lock.json`, do you mean deleting it and re-creating it?

Reply:
`npm-check-updates` has the [doctor](https://github.com/raineorshine/npm-check-updates#doctor-mode) functionality which is super powerful for safely updating every package to latest

Yep, that's what I meant!

Reply:
ooh I see! So we should definitely try `npm-check-updates`'s doctor command then! Thanks for the research 🚀 !

Reply:
I'm going to make a decision here that we can revisit in the future: as part of monthly package updates, if there are breaking changes that occur due to major package updates, we should simply address them right away as part of housekeeping and ensuring that we don't continue to rack up debt. The failure I want to avoid is that we continuously deprioritize package updates because they are more work. Let's run into the case where the package update takes a ton of time away from a sprint, before installing the instructions to create a ticket and prioritize.

Reply:
I updated the prose! Please re-review when you have a chance.